File: pkg/planner/core/rule_join_elimination.go
// Copyright 2018 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package core

import (
	"context"

	"github.com/pingcap/tidb/pkg/expression"
	"github.com/pingcap/tidb/pkg/parser/ast"
	"github.com/pingcap/tidb/pkg/parser/mysql"
	"github.com/pingcap/tidb/pkg/planner/core/base"
	"github.com/pingcap/tidb/pkg/planner/core/operator/logicalop"
	ruleutil "github.com/pingcap/tidb/pkg/planner/core/rule/util"
	coreusage "github.com/pingcap/tidb/pkg/planner/util/coreusage"
	"github.com/pingcap/tidb/pkg/sessionctx/vardef"
	"github.com/pingcap/tidb/pkg/util/intset"
)

// OuterJoinEliminator is used to eliminate outer join.
type OuterJoinEliminator struct {
}

// appendUniqueCorrelatedCols appends correlated columns to the target slice, avoiding duplicates.
func appendUniqueCorrelatedCols(target []*expression.Column, corCols []*expression.CorrelatedColumn) []*expression.Column {
	if len(corCols) == 0 {
		return target
	}
	seen := make(map[int64]struct{})
	for _, cc := range corCols {
		uid := cc.Column.UniqueID
		if _, ok := seen[uid]; ok {
			continue
		}
		seen[uid] = struct{}{}
		target = append(target, &cc.Column)
	}
	return target
}

// tryToEliminateOuterJoin will eliminate outer join plan base on the following rules
//  1. outer join elimination: For example left outer join, if the parent doesn't use the
//     columns from right table and the join key of right table(the inner table) is a unique
//     key of the right table. the left outer join can be eliminated.
//  2. outer join elimination with duplicate agnostic aggregate functions: For example left outer join.
//     If the parent only use the columns from left table with 'distinct' label. The left outer join can
//     be eliminated.
func (o *OuterJoinEliminator) tryToEliminateOuterJoin(p *logicalop.LogicalJoin, aggCols []*expression.Column, parentCols []*expression.Column) (base.LogicalPlan, bool, error) {
	var innerChildIdx int
	switch p.JoinType {
	case base.LeftOuterJoin:
		innerChildIdx = 1
	case base.RightOuterJoin:
		innerChildIdx = 0
	default:
		return p, false, nil
	}

	outerPlan := p.Children()[1^innerChildIdx]
	innerPlan := p.Children()[innerChildIdx]

	// in case of count(*) FROM R LOJ S, the parentCols is empty, but
	// still need to proceed to check whether we can eliminate outer join.
	// In fact, we only care about whether there is any column from inner
	// table, if there is none, we are good.
	if len(parentCols) > 0 {
		outerUniqueIDs := intset.NewFastIntSet()
		for _, outerCol := range outerPlan.Schema().Columns {
			outerUniqueIDs.Insert(int(outerCol.UniqueID))
		}
		matched := ruleutil.IsColsAllFromOuterTable(parentCols, &outerUniqueIDs)
		if !matched {
			return p, false, nil
		}
	}

	if len(aggCols) > 0 {
		innerUniqueIDs := intset.NewFastIntSet()
		for _, innerCol := range innerPlan.Schema().Columns {
			innerUniqueIDs.Insert(int(innerCol.UniqueID))
		}
		// Check if any column is from the inner table.
		// If any column is from the inner table, we cannot eliminate the outer join.
		innerFound := ruleutil.IsColFromInnerTable(aggCols, &innerUniqueIDs)
		if !innerFound {
			return outerPlan, true, nil
		}
	}
	// outer join elimination without duplicate agnostic aggregate functions
	innerJoinKeys, innerNullEQKeys := o.extractInnerJoinKeys(p, innerChildIdx)
	contain, err := o.isInnerJoinKeysContainUniqueKey(innerPlan, innerJoinKeys, innerNullEQKeys)
	if err != nil {
		return p, false, err
	}
	if contain {
		return outerPlan, true, nil
	}
	contain, err = o.isInnerJoinKeysContainIndex(innerPlan, innerJoinKeys, innerNullEQKeys)
	if err != nil {
		return p, false, err
	}
	if contain {
		return outerPlan, true, nil
	}

	return p, false, nil
}

// extract join keys as a schema for inner child of a outer join, and record which inner join keys use NullEQ (<=>).
func (*OuterJoinEliminator) extractInnerJoinKeys(join *logicalop.LogicalJoin, innerChildIdx int) (*expression.Schema, intset.FastIntSet) {
	joinKeys := make([]*expression.Column, 0, len(join.EqualConditions))
	innerNullEQKeys := intset.NewFastIntSet()
	for _, eqCond := range join.EqualConditions {
		innerKey := eqCond.GetArgs()[innerChildIdx].(*expression.Column)
		joinKeys = append(joinKeys, innerKey)
		if eqCond.FuncName.L == ast.NullEQ {
			innerNullEQKeys.Insert(int(innerKey.UniqueID))
		}
	}
	return expression.NewSchema(joinKeys...), innerNullEQKeys
}

// check whether one of unique keys sets is contained by inner join keys
func (*OuterJoinEliminator) isInnerJoinKeysContainUniqueKey(innerPlan base.LogicalPlan, joinKeys *expression.Schema, innerNullEQKeys intset.FastIntSet) (bool, error) {
	for _, keyInfo := range innerPlan.Schema().PKOrUK {
		joinKeysContainKeyInfo := true
		for _, col := range keyInfo {
			if !joinKeys.Contains(col) {
				joinKeysContainKeyInfo = false
				break
			}
		}
		if joinKeysContainKeyInfo {
			return true, nil
		}
	}
	for _, keyInfo := range innerPlan.Schema().NullableUK {
		joinKeysContainKeyInfo := true
		for _, col := range keyInfo {
			if !joinKeys.Contains(col) {
				joinKeysContainKeyInfo = false
				break
			}
			if innerNullEQKeys.Has(int(col.UniqueID)) {
				joinKeysContainKeyInfo = false
				break
			}
		}
		if joinKeysContainKeyInfo {
			return true, nil
		}
	}
	return false, nil
}

// check whether one of index sets is contained by inner join index
func (*OuterJoinEliminator) isInnerJoinKeysContainIndex(innerPlan base.LogicalPlan, joinKeys *expression.Schema, innerNullEQKeys intset.FastIntSet) (bool, error) {
	ds, ok := innerPlan.(*logicalop.DataSource)
	if !ok {
		return false, nil
	}
	for _, path := range ds.AllPossibleAccessPaths {
		if path.IsIntHandlePath || !path.Index.Unique || len(path.IdxCols) == 0 {
			continue
		}
		joinKeysContainIndex := true
		for _, idxCol := range path.IdxCols {
			if !joinKeys.Contains(idxCol) {
				joinKeysContainIndex = false
				break
			}
			if !mysql.HasNotNullFlag(idxCol.RetType.GetFlag()) && innerNullEQKeys.Has(int(idxCol.UniqueID)) {
				joinKeysContainIndex = false
				break
			}
		}
		if joinKeysContainIndex {
			return true, nil
		}
	}
	return false, nil
}

func (o *OuterJoinEliminator) doOptimize(p base.LogicalPlan, aggCols []*expression.Column, parentCols []*expression.Column) (base.LogicalPlan, error) {
	// CTE's logical optimization is independent.
	if _, ok := p.(*logicalop.LogicalCTE); ok {
		return p, nil
	}
	var err error
	var isEliminated bool
	for join, isJoin := p.(*logicalop.LogicalJoin); isJoin; join, isJoin = p.(*logicalop.LogicalJoin) {
		p, isEliminated, err = o.tryToEliminateOuterJoin(join, aggCols, parentCols)
		if err != nil {
			return p, err
		}
		if !isEliminated {
			break
		}
	}

	switch x := p.(type) {
	case *logicalop.LogicalJoin:
		parentCols = parentCols[:0]
		parentCols = append(parentCols, x.Schema().Columns...)
		if x.JoinType == base.LeftOuterJoin || x.JoinType == base.RightOuterJoin {
			outerIdx := 0
			if x.JoinType == base.RightOuterJoin {
				outerIdx = 1
			}
			outerSchema := x.Children()[outerIdx].Schema()
			if join, ok := x.Children()[outerIdx].(*logicalop.LogicalJoin); ok && join.FullSchema != nil {
				outerSchema = join.FullSchema
			}
			condCols := make([]*expression.Column, 0)
			condCols = append(condCols, expression.ExtractColumnsFromExpressions(x.OtherConditions, nil)...)
			condCols = append(condCols, expression.ExtractColumnsFromExpressions(x.LeftConditions, nil)...)
			condCols = append(condCols, expression.ExtractColumnsFromExpressions(x.RightConditions, nil)...)
			condCols = append(condCols, expression.ExtractColumnsFromExpressions(expression.ScalarFuncs2Exprs(x.EqualConditions), nil)...)
			condCols = append(condCols, expression.ExtractColumnsFromExpressions(expression.ScalarFuncs2Exprs(x.NAEQConditions), nil)...)
			for _, col := range condCols {
				if outerSchema != nil && outerSchema.Contains(col) {
					parentCols = append(parentCols, col)
				}
			}
		}
	case *logicalop.LogicalApply:
		// TODO: this is tied to variable tidb_opt_enable_no_decorrelate_in_select
		// to enable outer join elimination when correlated subqueries exist in the select
		// list. Future enhancement can remove the variable check.
		x.SCtx().GetSessionVars().RecordRelevantOptVar(vardef.TiDBOptEnableNoDecorrelateInSelect)
		if x.SCtx().GetSessionVars().EnableNoDecorrelateInSelect {
			// For Apply, only columns from the left child are visible to the parent at this layer.
			// Filter incoming parentCols to left child's schema and also include correlated columns
			// from the right child that reference the left schema (required by subqueries).
			leftSchema := x.Children()[0].Schema()
			filtered := make([]*expression.Column, 0, len(parentCols))
			for _, col := range parentCols {
				if leftSchema.Contains(col) {
					filtered = append(filtered, col)
				}
			}
			// Add correlated columns from the right child that map to the left schema
			corCols := coreusage.ExtractCorColumnsBySchema4LogicalPlan(x.Children()[1], leftSchema)
			filtered = appendUniqueCorrelatedCols(filtered, corCols)
			parentCols = filtered
		} else {
			parentCols = append(parentCols[:0], p.Schema().Columns...)
		}
	case *logicalop.LogicalProjection:
		parentCols = parentCols[:0]
		for _, expr := range x.Exprs {
			parentCols = append(parentCols, expression.ExtractColumns(expr)...)
		}
		// Include columns required by subqueries (appear as correlated columns in child subtree)
		// TODO: this is tied to variable tidb_opt_enable_no_decorrelate_in_select
		// to enable outer join elimination when correlated subqueries exist in the select
		// list. Future enhancement can remove the variable check.
		if len(x.Children()) > 0 {
			x.SCtx().GetSessionVars().RecordRelevantOptVar(vardef.TiDBOptEnableNoDecorrelateInSelect)
			if x.SCtx().GetSessionVars().EnableNoDecorrelateInSelect {
				corCols := coreusage.ExtractCorrelatedCols4LogicalPlan(x.Children()[0])
				parentCols = appendUniqueCorrelatedCols(parentCols, corCols)
			}
		}
	case *logicalop.LogicalAggregation:
		parentCols = parentCols[:0]
		for _, groupByItem := range x.GroupByItems {
			parentCols = append(parentCols, expression.ExtractColumns(groupByItem)...)
		}
		for _, aggDesc := range x.AggFuncs {
			for _, expr := range aggDesc.Args {
				parentCols = append(parentCols, expression.ExtractColumns(expr)...)
			}
			for _, byItem := range aggDesc.OrderByItems {
				parentCols = append(parentCols, expression.ExtractColumns(byItem.Expr)...)
			}
		}
	default:
		parentCols = append(parentCols[:0], p.Schema().Columns...)
	}

	if ok, newCols := logicalop.GetDupAgnosticAggCols(p, aggCols); ok {
		aggCols = newCols
	}

	for i, child := range p.Children() {
		newChild, err := o.doOptimize(child, aggCols, parentCols)
		if err != nil {
			return nil, err
		}
		p.SetChild(i, newChild)
	}
	return p, nil
}

// Optimize implements base.LogicalOptRule.<0th> interface.
func (o *OuterJoinEliminator) Optimize(_ context.Context, p base.LogicalPlan) (base.LogicalPlan, bool, error) {
	planChanged := false
	p, err := o.doOptimize(p, nil, nil)
	return p, planChanged, err
}

// Name implements base.LogicalOptRule.<1st> interface.
func (*OuterJoinEliminator) Name() string {
	return "outer_join_eliminate"
}

File: rust/crates/tidb-planner/src/logical/rule_join_elimination.rs
// Copyright 2026 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Go `pkg/planner/core/rule_join_elimination.go`.

use std::collections::HashSet;

use tidb_datatype::{Datum, FieldTypeFlags};
use tidb_expr::aggregation::names;
use tidb_expr::column::Column;
use tidb_expr::constant::Constant;
use tidb_expr::expr_util::extract::find_upper_bound;
use tidb_expr::expression::Expression;
use tidb_expr::schema::Schema;
use tidb_expr::simple_expr::extract_columns;

use crate::access_path::DataSourceAccessPath;
use crate::expression_rewriter::{
    extract_cor_columns_by_schema_4_logical_plan, extract_correlated_cols_4_logical_plan,
};
use crate::find_best_task::LogicalJoinType;

use super::rule::{LogicalOptRule, RuleContext};
use super::window::{BoundType, FrameType};
use super::{
    BaseLogicalPlan, LogicalAggregation, LogicalJoin, LogicalPlan, LogicalProjection, PlanError,
};

/// Go `OuterJoinEliminator`.
pub struct OuterJoinEliminator;

fn columns_from(expressions: impl IntoIterator<Item = Expression>) -> Vec<Column> {
    expressions
        .into_iter()
        .flat_map(|expression| extract_columns(&expression))
        .collect()
}

fn append_unique(target: &mut Vec<Column>, columns: impl IntoIterator<Item = Column>) {
    let mut seen = HashSet::new();
    for column in columns {
        if seen.insert(column.unique_id) {
            target.push(column);
        }
    }
}

fn null_extended_projection(
    context: &RuleContext<'_>,
    join: &LogicalJoin,
    outer: LogicalPlan,
) -> LogicalPlan {
    let Some(join_schema) = join.base.base.schema().cloned() else {
        return outer;
    };
    let Some(outer_schema) = outer.schema() else {
        return outer;
    };
    let mut all_from_outer = true;
    let expressions = join_schema
        .columns
        .iter()
        .map(|column| {
            if outer_schema.contains(column) {
                Expression::Column(column.clone())
            } else {
                all_from_outer = false;
                let mut field_type = column
                    .ret_type
                    .clone()
                    .expect("logical schema column has RetType");
                field_type.del_flags(FieldTypeFlags::NOT_NULL);
                Expression::Constant(Constant::new(Datum::Null, field_type))
            }
        })
        .collect();
    if all_from_outer {
        return outer;
    }
    let mut base = BaseLogicalPlan::new(
        context.allocator,
        LogicalProjection::TYPE,
        join.base.base.query_block_offset(),
    );
    base.base.set_schema(Some(join_schema));
    base.base
        .set_output_names(join.base.base.output_names().to_vec());
    base.set_children(vec![outer]);
    LogicalPlan::Projection(LogicalProjection::new(base, expressions))
}

fn extract_inner_join_keys(join: &LogicalJoin, inner_index: usize) -> (Schema, HashSet<i64>) {
    let mut keys = Vec::with_capacity(join.equal_conditions.len());
    let mut null_eq = HashSet::new();
    let inner_schema = join
        .base
        .children()
        .get(inner_index)
        .and_then(LogicalPlan::schema);
    for condition in &join.equal_conditions {
        let Some(column) = condition.args.iter().find_map(|argument| {
            let Expression::Column(column) = argument else {
                return None;
            };
            inner_schema
                .is_some_and(|schema| schema.contains(column))
                .then_some(column)
        }) else {
            continue;
        };
        keys.push(column.clone());
        if condition.func_name.lowercase() == "nulleq" {
            null_eq.insert(column.unique_id);
        }
    }
    (Schema::new(keys), null_eq)
}

fn is_col_eq_one(expression: &Expression, target: &Column) -> bool {
    let Expression::ScalarFunction(function) = expression else {
        return false;
    };
    if function.func_name.lowercase() != "eq" || function.args.len() != 2 {
        return false;
    }
    let matches = |column: &Expression, constant: &Expression| {
        matches!(column, Expression::Column(column) if column.unique_id == target.unique_id)
            && matches!(constant, Expression::Constant(constant) if constant.value == Datum::Int(1))
    };
    matches(&function.args[0], &function.args[1]) || matches(&function.args[1], &function.args[0])
}

fn row_number_partition_is_unique(inner: &LogicalPlan, join_keys: &Schema) -> bool {
    let LogicalPlan::Selection(selection) = inner else {
        return false;
    };
    if selection.conditions.is_empty() {
        return false;
    }
    let [LogicalPlan::Window(window)] = selection.base.children() else {
        return false;
    };
    if window.window_func_descs.len() != 1 || window.window_func_descs[0].base.name != "row_number"
    {
        return false;
    }
    let Some(frame) = &window.frame else {
        return false;
    };
    if frame.frame_type != FrameType::Rows
        || frame.start.as_ref().map(|bound| bound.bound_type) != Some(BoundType::CurrentRow)
        || frame.end.as_ref().map(|bound| bound.bound_type) != Some(BoundType::CurrentRow)
    {
        return false;
    }
    let Some(schema) = window.base.base.schema() else {
        return false;
    };
    let [row_number] = window.get_window_result_columns(schema) else {
        return false;
    };
    let bounded = selection.conditions.iter().any(|condition| {
        is_col_eq_one(condition, row_number)
            || find_upper_bound(condition).is_some_and(|(column, upper)| {
                column.unique_id == row_number.unique_id && upper <= 1
            })
    });
    bounded
        && window
            .get_partition_by_cols()
            .iter()
            .all(|column| join_keys.contains(column))
}

fn join_keys_contain_schema_key(
    inner: &LogicalPlan,
    join_keys: &Schema,
    null_eq: &HashSet<i64>,
) -> bool {
    if row_number_partition_is_unique(inner, join_keys) {
        return true;
    }
    let Some(schema) = inner.schema() else {
        return false;
    };
    if schema
        .pk_or_uk
        .iter()
        .any(|key| key.iter().all(|column| join_keys.contains(column)))
    {
        return true;
    }
    schema.nullable_uk.iter().any(|key| {
        key.iter()
            .all(|column| join_keys.contains(column) && !null_eq.contains(&column.unique_id))
    })
}

fn join_keys_contain_unique_index(
    inner: &LogicalPlan,
    join_keys: &Schema,
    null_eq: &HashSet<i64>,
) -> bool {
    let LogicalPlan::DataSource(source) = inner else {
        return false;
    };
    source.all_possible_access_paths.iter().any(|path| {
        let DataSourceAccessPath::Index(path) = path else {
            return false;
        };
        let index_id = path.candidate().index_id;
        let Some(index) = source
            .indexes
            .iter()
            .find(|index| index.id == index_id && index.unique && !index.columns.is_empty())
        else {
            return false;
        };
        index.columns.iter().all(|index_column| {
            let Some(column) = source.table_columns.get(index_column.offset) else {
                return false;
            };
            join_keys.contains(column)
                && (column
                    .ret_type
                    .as_ref()
                    .is_some_and(|field_type| field_type.has_flag(FieldTypeFlags::NOT_NULL))
                    || !null_eq.contains(&column.unique_id))
        })
    })
}

fn try_eliminate(
    context: &RuleContext<'_>,
    mut join: LogicalJoin,
    agg_columns: &[Column],
    parent_columns: &[Column],
) -> (LogicalPlan, bool) {
    let inner_index = match join.join_type {
        LogicalJoinType::LeftOuter => 1,
        LogicalJoinType::RightOuter => 0,
        _ => return (LogicalPlan::Join(join), false),
    };
    let outer_index = 1 ^ inner_index;
    let children = join.base.children();
    let Some(inner) = children.get(inner_index) else {
        return (LogicalPlan::Join(join), false);
    };
    let Some(outer) = children.get(outer_index) else {
        return (LogicalPlan::Join(join), false);
    };
    if matches!(inner, LogicalPlan::TableDual(dual) if dual.row_count == 0) {
        let mut children = join.base.take_children();
        let outer = children.swap_remove(outer_index);
        return (null_extended_projection(context, &join, outer), true);
    }
    let Some(outer_schema) = outer.schema() else {
        return (LogicalPlan::Join(join), false);
    };
    if parent_columns
        .iter()
        .any(|column| !outer_schema.contains(column))
    {
        return (LogicalPlan::Join(join), false);
    }
    if !agg_columns.is_empty() {
        let Some(inner_schema) = inner.schema() else {
            return (LogicalPlan::Join(join), false);
        };
        if agg_columns
            .iter()
            .all(|column| !inner_schema.contains(column))
        {
            let mut children = join.base.take_children();
            return (children.swap_remove(outer_index), true);
        }
    }
    let (join_keys, null_eq) = extract_inner_join_keys(&join, inner_index);
    if join_keys_contain_schema_key(inner, &join_keys, &null_eq)
        || join_keys_contain_unique_index(inner, &join_keys, &null_eq)
    {
        let mut children = join.base.take_children();
        return (children.swap_remove(outer_index), true);
    }
    (LogicalPlan::Join(join), false)
}

fn duplicate_agnostic_columns(aggregation: &LogicalAggregation) -> Vec<Column> {
    let mut columns = Vec::new();
    for function in &aggregation.agg_funcs {
        if !function.has_distinct
            && !matches!(
                function.name(),
                names::FIRST_ROW | names::MAX | names::MIN | names::APPROX_COUNT_DISTINCT
            )
        {
            return Vec::new();
        }
        for argument in function.args() {
            columns.extend(extract_columns(argument));
        }
    }
    columns
}

fn required_columns(
    context: &RuleContext<'_>,
    plan: &LogicalPlan,
    incoming: &[Column],
) -> Vec<Column> {
    match plan {
        LogicalPlan::Apply(apply) => {
            if context.enable_no_decorrelate_in_select {
                let Some(left) = apply.join.base.children().first() else {
                    return Vec::new();
                };
                let Some(left_schema) = left.schema() else {
                    return Vec::new();
                };
                let mut columns = incoming
                    .iter()
                    .filter(|column| left_schema.contains(column))
                    .cloned()
                    .collect();
                if let Some(right) = apply.join.base.children().get(1) {
                    append_unique(
                        &mut columns,
                        extract_cor_columns_by_schema_4_logical_plan(right, left_schema)
                            .into_iter()
                            .map(|column| column.column),
                    );
                }
                columns
            } else {
                plan.schema()
                    .map_or_else(Vec::new, |schema| schema.columns.clone())
            }
        }
        LogicalPlan::Projection(projection) => {
            let mut columns = columns_from(projection.exprs.clone());
            if context.enable_no_decorrelate_in_select {
                if let Some(child) = projection.base.children().first() {
                    append_unique(
                        &mut columns,
                        extract_correlated_cols_4_logical_plan(child)
                            .into_iter()
                            .map(|column| column.column),
                    );
                }
            }
            columns
        }
        LogicalPlan::Aggregation(aggregation) => {
            let mut columns = columns_from(aggregation.group_by_items.clone());
            for function in &aggregation.agg_funcs {
                columns.extend(columns_from(function.args().to_vec()));
                columns.extend(columns_from(
                    function.order_by_items.iter().map(|item| item.expr.clone()),
                ));
            }
            columns
        }
        LogicalPlan::Join(join) => {
            let mut columns = plan
                .schema()
                .map_or_else(Vec::new, |schema| schema.columns.clone());
            columns.extend(columns_from(
                join.equal_conditions
                    .iter()
                    .cloned()
                    .map(Expression::ScalarFunction),
            ));
            columns.extend(columns_from(join.left_conditions.clone()));
            columns.extend(columns_from(join.right_conditions.clone()));
            columns.extend(columns_from(join.other_conditions.clone()));
            columns.extend(columns_from(
                join.na_eq_conditions
                    .iter()
                    .cloned()
                    .map(Expression::ScalarFunction),
            ));
            columns
        }
        _ => plan
            .schema()
            .map_or_else(Vec::new, |schema| schema.columns.clone()),
    }
}

fn optimize_recursive(
    context: &RuleContext<'_>,
    mut plan: LogicalPlan,
    agg_columns: Vec<Column>,
    parent_columns: Vec<Column>,
) -> LogicalPlan {
    if matches!(plan, LogicalPlan::CTE(_)) {
        return plan;
    }
    loop {
        let LogicalPlan::Join(join) = plan else {
            break;
        };
        let (rewritten, eliminated) = try_eliminate(context, join, &agg_columns, &parent_columns);
        plan = rewritten;
        if !eliminated {
            break;
        }
    }

    let child_parent_columns = required_columns(context, &plan, &parent_columns);
    let child_agg_columns = match &plan {
        LogicalPlan::Aggregation(aggregation) => duplicate_agnostic_columns(aggregation),
        _ => agg_columns,
    };
    let children = plan
        .base_mut()
        .take_children()
        .into_iter()
        .map(|child| {
            optimize_recursive(
                context,
                child,
                child_agg_columns.clone(),
                child_parent_columns.clone(),
            )
        })
        .collect();
    plan.base_mut().set_children(children);
    plan
}

impl LogicalOptRule for OuterJoinEliminator {
    fn optimize(
        &self,
        context: &RuleContext<'_>,
        plan: LogicalPlan,
    ) -> Result<(LogicalPlan, bool), (LogicalPlan, PlanError)> {
        Ok((
            optimize_recursive(context, plan, Vec::new(), Vec::new()),
            false,
        ))
    }

    fn name(&self) -> &'static str {
        "outer_join_eliminate"
    }
}
